### ISSUE TITLE
Some issue after “Decontainerize non-scalar symbols upon import” commit

### ISSUE THREAD

**Issue by AlexDaniel** [rakudo/rakudo#3012]
Pastebin::Shadowcat no longer passes its test.

Bisected: https://github.com/rakudo/rakudo/commit/3af3bde70eaed13b235bc55c54cf0ca604be6fcf

Error:
```
Cannot invoke this object (REPR: Null; VMNull)
  in sub decode at /root/Blin/installed/HTML::Entity_*/sources/E6858CA39AB81D28F90ED1A7D6D1DC66D0E9FAA5 (HTML::Entity) line 15
  in method fetch at /root/Blin/data/zef-data/store/perl6-Pastebin-Shadowcat.git/7e8769e51fd27618681b2d6be623c6c64c996a9b/lib/Pastebin/Shadowcat.pm6 (Pastebin::Shadowcat) line 30
  in block <unit> at t/01-paste.t line 12
```

Related pieces of code:
* https://github.com/perl6-community-modules/perl6-Pastebin-Shadowcat/blob/7e8769e51fd27618681b2d6be623c6c64c996a9b/lib/Pastebin/Shadowcat.pm6#L31
* https://github.com/Mouq/HTML-Entity/blob/57931c74f08344fe06459368d80f2b3a16cc5dee/lib/HTML/Entity.pm6#L25

Seems like something is not right in rakudo and the module is not at fault.

Ping @vrurg.

**Comment by vrurg** [comment]
I have golfed it down to [this](https://gist.github.com/vrurg/6f7d4940a62648e76b28a429c39d1113). The point is that my commit is unrelated in this case. I have reverted it on a local branch and the outcome is still the same – `Cannot invoke ...`. My commit could have resulted in more predictable behavior of the inital problem.

Debug printing of `&test-test.WHAT` reveals that within `Baz` module it's `Sub`; whereas in `cannot_invoke.p6` it turns into `Callable`.

What could possibly be the good outcome of this is that I probably know where to look for the source of containerization problem which I was trying to fix with `3af3bde`.

**Comment by vrurg** [comment]
Final for today. The problem is in binding. `is export` trait is applied before binding takes place. Correspondingly, the `Scalar` container with `Callable` type object gets exported. Afterwards `&test-test` is bound to `&test` and by this moment it's no more linked to the exported `Scalar`.

A solution to this is to use 'constant' instead of binding. The semantics is basically the same but constants are handled differently and are not affected by this problem. Effectively, use of binding with `is export` must result in either a error or a warning. As a possibility, I could look into how constants are handled and possibly do the same for variables if this is possible at all.

**Comment by vrurg** [comment]
More generic golfing of the problem:

```
module Foo { my $foo = 42; our $FOO is export := $foo }; import Foo; say $FOO; # (Any)
```

Should it work whatsoever? Assignment works because it's the same `Scalar` which is exported and which is being assigned into.

**Comment by vrurg** [comment]
Merged #3016 which fixes this case. `Pastebin::Shadowcat` passes `prove6 -l` for me now.

If no other objections, the ticket could be closed.
